src/webgpu/shader/types.ts

```diff
@@ -36,3 +42,146 @@ export const kMatrixContainerTypeInfo = /* prettier-ignore */ {
 } as const;
 /** List of all matNxN<> container types. */
 export const kMatrixContainerTypes = keysOf(kMatrixContainerTypeInfo);
+
+/**
+ * Generate a bunch types (vec, mat, sized/unsized array) for testing.
+ */
+export function* generateTypes({
+  storageClass,
+  baseType,
+  containerType,
+  supportsAtomics = false,
+}: {
+  storageClass: string;
+  /** Base scalar type (i32/u32/f32/bool). */
+  baseType: ScalarType;
+  /** Container type (scalar/vector/matrix/array) */
+  containerType: ContainerType;
+  /** Whether to wrap the baseType in `atomic<>`. */
+  supportsAtomics?: boolean;
+}) {
+  const scalarInfo = kScalarTypeInfo[baseType];
+  if (supportsAtomics) {
+    assert(scalarInfo.supportsAtomics, 'type does not support atomics');
+  }
+  const scalarType = supportsAtomics ? `atomic<${baseType}>` : baseType;
+
+  // Storage and uniform require host-sharable types.
+  if (storageClass === 'storage' || storageClass === 'uniform') {
+    assert(isHostSharable(baseType), 'type ' + baseType.toString() + ' is not host sharable');
+  }
+
+  // Scalar types
+  if (containerType === 'scalar') {
+    yield {
+      type: `${scalarType}`,
```

Comment:
maybe "type" can be an input parameter to this function. I can see lots of tests wanting to use "type" to mean something else

Reply:
I think "type" is the same as baseType here. Or maybe I don't know what you mean by "type".
Could you give an example? Or refer me to one the tests using "type".

Reply:
Sorry, I wasn't very clear. Also though, this is a small usability nit. I think it's OK if we leave it as-is for now.

At other times, "type" is:
"type: `array<${scalarType},${kArrayLength}>`"

If I understand correctly, `generateTypes` returns:
```
Generator<{
  type: string,
  _kTypeInfo: {
    elementBaseType: string,
    // other type info..
  }
}>
```

It could cause a conflict if the place where the helper is used wants to use "type" for something else:

```
const enum OtherType {
  Small, Medium, Big
}

.params(u => 
  u
    .combine('type', [OtherType.Small, OtherType.Medium, OtherType.Big])
    .beginSubcases()
    .expandWithParams(generateTypes)
)
```

This would be a problem because there are two params with the same name "type".

We could *maybe* make it look like:
```
  u
    .combine('type', [OtherType.Small, OtherType.Medium, OtherType.Big])
    .beginSubcases()
    .expandWithParams(generateTypes.bind('wgslType'))
    .expandWithParams(generateTypes.bind('otherWgslType'))
)
```
so it names the param "wgslType" instead.

Reply:
Let's leave this as-is. I'll fix it in my next CL.